### PR TITLE
feat(csms): add a resource to manage CSMS secrets

### LINKED ISSUE #1858: I want to add a resource for managing the secrets of Data Encryption Workshop
I want to create, modify and delete secrets of Data Encryption Workshop.
The resource name: `huaweicloud_dew_secret`

### PR DESCRIPTION
**Pull request by chengxiangdong**
**What this PR does / why we need it**:

Add a resource to manage CSMS secrets.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1858

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

### Resource
```
make testacc TEST='./huaweicloud/services/acceptance/dew' TESTARGS='-run TestAccDewCsmsSecret_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dew -v -run TestAccDewCsmsSecret_basic -timeout 360m -parallel 4
=== RUN   TestAccDewCsmsSecret_basic
=== PAUSE TestAccDewCsmsSecret_basic
=== CONT  TestAccDewCsmsSecret_basic
--- PASS: TestAccDewCsmsSecret_basic (64.28s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dew       64.359s
```

### Endpoint
```
make testacc TEST='./huaweicloud/services/acceptance/dew' TESTARGS='-run TestAccDewCsmsSecret_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dew -v -run TestAccDewCsmsSecret_basic -timeout 360m -parallel 4
=== RUN   TestAccDewCsmsSecret_basic
=== PAUSE TestAccDewCsmsSecret_basic
=== CONT  TestAccDewCsmsSecret_basic
--- PASS: TestAccDewCsmsSecret_basic (64.28s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dew       64.359s
cheng@ecs-chengxiangdong:/mnt/d/code/terraform-provider-huaweicloud$ make testacc TEST='./huaweicloud' TESTARGS='-run TestAccServiceEndpoints_Security'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud -v -run TestAccServiceEndpoints_Security -timeout 360m -parallel 4
=== RUN   TestAccServiceEndpoints_Security
...
...
...
    endpoints_test.go:305: anti-ddos endpoint:   https://antiddos.cn-north-4.myhuaweicloud.com/v1/0970dd7a1300f5672ff2c003c60ae115/
    endpoints_test.go:317: KMS(v1.0) endpoint:   https://kms.cn-north-4.myhuaweicloud.com/v1.0/
    endpoints_test.go:329: KMS(v1) endpoint:     https://kms.cn-north-4.myhuaweicloud.com/v1/0970dd7a1300f5672ff2c003c60ae115/
    endpoints_test.go:341: SCM endpoint:         https://scm.cn-north-4.myhuaweicloud.com/v3/
    endpoints_test.go:353: WAF endpoint:         https://waf.cn-north-4.myhuaweicloud.com/v1/0970dd7a1300f5672ff2c003c60ae115/waf/
    endpoints_test.go:365: WAF dedicated endpoint:       https://waf.cn-north-4.myhuaweicloud.com/v1/0970dd7a1300f5672ff2c003c60ae115/premium-waf/
--- PASS: TestAccServiceEndpoints_Security (1.42s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud       1.549s

```

### CODE REVIEW

File: docs/resources/csms_secret.md [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+  Only digits, letters, underscores(_), hyphens(-) and dots(.) are allowed.
+
+* `secret_text` - (Required, String) The plaintext of a secret in text format. The maximum size is 32 KB.
+
+  -> **NOTE:** The `secret_text` is sensitive and in the state file we store its hash.
+
+* `kms_key_id` - (Optional, String) The ID of the KMS key used to encrypt secrets.
+  If this parameter is not specified when creating the secret, the default master key csms/default will be used.
+  The default key is automatically created by the CSMS.
+  Use this data source `huaweicloud_kms_key` to get the KMS key.
```
Comment (niuzhenguo) [inline docs/resources/csms_secret.md]: better to make `haweicloud_kms_key` as a link to the resource doc.

File: docs/resources/csms_secret.md [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+* `id` - The resource ID which is constructed from the secret ID and name, separated by slashes.
+
+* `secret_id` - The secret ID in UUID format.
+
+* `latest_version` - The latest version id.
+
+* `state` - The CSMS secret status. Values can be: ENABLED, DISABLED, PENDING_DELETE and FROZEN.
+
+* `create_time` - Time when the CSMS secrets created, in UTC format.
```
Comment (niuzhenguo) [inline docs/resources/csms_secret.md:69]: `create_time` or `created_time`? do we have a recomended/standard name for this?

File: docs/resources/csms_secret.md [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+## Import
+
+CSMS secret can be imported using the ID and name of secret, separated by a slash, e.g.
```
Comment (niuzhenguo) [inline docs/resources/csms_secret.md]: name of `the` secret

File: huaweicloud/provider.go [hunk 4]
```diff
@@ -608,6 +609,7 @@ func Provider() *schema.Provider {
 			"huaweicloud_waf_instance_group":              waf.ResourceWafInstanceGroup(),
 			"huaweicloud_waf_instance_group_associate":    waf.ResourceWafInstGroupAssociate(),
 			"huaweicloud_waf_reference_table":             waf.ResourceWafReferenceTableV1(),
+			"huaweicloud_csms_secret":                     dew.ResourceCsmsSecret(),
```
Comment (niuzhenguo) [inline huaweicloud/provider.go]: please move this above, we make this list in alphabetical order.

File: docs/resources/csms_secret.md (hunk 2 above)
Comment (chengxiangdong) [reply docs/resources/csms_secret.md:69]: using `created_time` is recommended in the normative doc, but we always use `create_time`

File: huaweicloud/provider.go (hunk 4 above)
Comment (chengxiangdong) [reply huaweicloud/provider.go]: fixed

File: docs/resources/csms_secret.md (hunk 1 above)
Comment (chengxiangdong) [reply docs/resources/csms_secret.md]: ok

File: huaweicloud/services/dew/resource_huaweicloud_csms_secret.go [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+		return fmtp.DiagErrorf("failed to delete CSMS secret : %s", err)
+	}
+	d.SetId("")
+	return nil
+}
+
+func resourceCsmsSecretImport(ctx context.Context, d *schema.ResourceData, _ interface{}) ([]*schema.ResourceData,
+	error) {
+	id, name := parseID(d.Id())
+	if len(id) == 0 {
```
Comment (hmbig2) [inline huaweicloud/services/dew/resource_huaweicloud_csms_secret.go]: name is required , please add check

File: huaweicloud/services/dew/resource_huaweicloud_csms_secret.go [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+				ValidateFunc: validation.StringMatch(regexp.MustCompile(`^[\w-\.]{1,64}$`),
+					"The maximum length is 64 characters. "+
+						"Only letters, digits, underscores (_) hyphens (-) and dots (.) are allowed."),
+			},
+			"secret_text": {
```
Comment (hmbig2) [inline huaweicloud/services/dew/resource_huaweicloud_csms_secret.go:54]: Binary strings and strings are quite different, why  merged

Comment (chengxiangdong) [reply huaweicloud/services/dew/resource_huaweicloud_csms_secret.go:54]: no merging, no support for binary data, same as the console

File: huaweicloud/services/dew/resource_huaweicloud_csms_secret.go (hunk 5 above)
Comment (chengxiangdong) [reply huaweicloud/services/dew/resource_huaweicloud_csms_secret.go]: parseID has do that

File: docs/resources/csms_secret.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
+---
+subcategory: "Data Encryption Workshop (DEW)"
+---
+
+# huaweicloud_csms_secret
+
+Manages CSMS(Cloud Secret Management Service) secrets within HuaweiCloud.
+
+## Example Usage
```
Comment (ShiChangkuo) [inline docs/resources/csms_secret.md:9]: suggest to add an example of `secret_text` in JSON format as same as the console.

File: docs/resources/csms_secret.md [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The resource ID which is constructed from the secret ID and name, separated by slashes.
+
+* `secret_id` - The secret ID in UUID format.
+
+* `latest_version` - The latest version id.
+
+* `state` - The CSMS secret status. Values can be: ENABLED, DISABLED, PENDING_DELETE and FROZEN.
```
Comment (ShiChangkuo) [inline docs/resources/csms_secret.md]: **status** is more common among huaweicloud provider.

File: huaweicloud/services/dew/resource_huaweicloud_csms_secret.go [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+	id := fmt.Sprintf("%s/%s", rst.ID, name)
+	d.SetId(id)
+
+	// Save tags
+	if t, ok := d.GetOk("tags"); ok {
+		tMaps := t.(map[string]interface{})
+		tagMaps := utils.ExpandResourceTags(tMaps)
+		err = tags.Create(client, serviceType, rst.ID, tagMaps).ExtractErr()
+		if err != nil {
+			logp.Printf("[WARN] Error add tags to CSMS secret :%s, err=%s", rst.ID, err)
```
Comment (ShiChangkuo) [inline huaweicloud/services/dew/resource_huaweicloud_csms_secret.go]: extra spaces

File: huaweicloud/services/dew/resource_huaweicloud_csms_secret.go [hunk 10]
```diff
@@ -0,0 +1,318 @@
...
+	)
+
+	// Query secret tags
+	if resourceTags, err := tags.Get(client, serviceType, id).Extract(); err == nil {
+		tagMap := utils.TagsToMap(resourceTags.Tags)
+		if err = d.Set("tags", tagMap); err != nil {
+			mErr = multierror.Append(mErr,
+				fmtp.Errorf("failed to set the tags of CSMS secret (%s) : %s", id, err),
+			)
+		}
```
Comment (ShiChangkuo) [inline huaweicloud/services/dew/resource_huaweicloud_csms_secret.go]: we can use multierror.Append directly, and judge the error at last

File: huaweicloud/services/dew/resource_huaweicloud_csms_secret.go (hunk 5 above)
Comment (ShiChangkuo) [inline huaweicloud/services/dew/resource_huaweicloud_csms_secret.go]: is `id == ""` more clear?

File: docs/resources/csms_secret.md (hunk 8 above)
Comment (chengxiangdong) [reply docs/resources/csms_secret.md]: fixed

File: docs/resources/csms_secret.md (hunk 7 above)
Comment (chengxiangdong) [reply docs/resources/csms_secret.md:9]: ok

File: huaweicloud/services/dew/resource_huaweicloud_csms_secret.go (hunk 5 above)
Comment (chengxiangdong) [reply huaweicloud/services/dew/resource_huaweicloud_csms_secret.go]: nice

**Pull request merged**